api/service/stagedsync/stage_blockhashes.go

```diff
@@ -0,0 +1,600 @@
+	if canRunInTurboMode {
+		var cacheHash []byte
+		if cacheHash, err = bh.getHashFromCache(currProgress + 1); err != nil {
+			utils.Logger().Info().
+				Msgf("[STAGED_SYNC] fetch cache progress for block hashes stage failed: %v", err)
+			return ErrFetchBlockHashProgressFail
+		} else {
+			if len(cacheHash[:]) > 0 {
+				// get blocks from cached db rather than calling peers, and update current progress
+				currProgress, startHash, err = bh.loadBlockHashesFromCache(s, currProgress, targetHeight, tx)
+				if err != nil {
+					utils.Logger().Info().
+						Msgf("[STAGED_SYNC] fetch cached block hashes failed: %v", err)
+					return ErrFetchCachedBlockHashFail
+				}
+			}
+		}
+	}
+
+	if currProgress >= targetHeight {
+		if canRunInTurboMode && currProgress < s.state.syncStatus.maxPeersHeight {
+			bh.configs.turboModeCh = make(chan struct{})
+			go bh.runBackgroundProcess(nil, s, isBeacon, currProgress, s.state.syncStatus.maxPeersHeight, startHash)
+		}
+		return nil
+	}
+
+	size := uint32(0)
+
+	fmt.Print("\033[s") // save the cursor position
+	startTime := time.Now()
+	startBlock := currProgress
+	for ok := true; ok; ok = currProgress < targetHeight {
+		size = uint32(targetHeight - currProgress)
+		if size > SyncLoopBatchSize {
+			size = SyncLoopBatchSize
+		}
+		// Gets consensus hashes.
+		if err := s.state.getConsensusHashes(startHash[:], size); err != nil {
+			return err
+		}
+		// selects the most common peer config based on their block hashes and doing the clean up
+		if err := s.state.syncConfig.GetBlockHashesConsensusAndCleanUp(); err != nil {
+			return err
+		}
+		// double check block hashes
+		if s.state.DoubleCheckBlockHashes {
+			invalidPeersMap, validBlockHashes, err := s.state.getInvalidPeersByBlockHashes(tx)
+			if err != nil {
+				return err
+			}
+			if validBlockHashes < int(size) {
+				return ErrNotEnoughBlockHashes
+			}
+			s.state.syncConfig.cleanUpInvalidPeers(invalidPeersMap)
+		}
+		// save the downloaded files to db
+		if currProgress, startHash, err = bh.saveDownloadedBlockHashes(s, currProgress, startHash, tx); err != nil {
+			return err
+		}
+
+		//calculating block speed
+		dt := time.Now().Sub(startTime).Seconds()
+		speed := float64(0)
+		if dt > 0 {
+			speed = float64(currProgress-startBlock) / dt
+		}
+		blockSpeed := fmt.Sprintf("%.2f", speed)
+
+		fmt.Print("\033[u\033[K") // restore the cursor position and clear the line
+		fmt.Println("downloading block hash progress:", currProgress, "/", targetHeight, "(", blockSpeed, "blocks/s", ")")
+	}
+
+	// continue downloading in background
+	if canRunInTurboMode && currProgress < s.state.syncStatus.maxPeersHeight {
+		bh.configs.turboModeCh = make(chan struct{})
+		go bh.runBackgroundProcess(nil, s, isBeacon, currProgress, s.state.syncStatus.maxPeersHeight, startHash)
+	}
+	return nil
+}
+
+func (bh *StageBlockHashes) runBackgroundProcess(tx kv.RwTx, s *StageState, isBeacon bool, startHeight uint64, targetHeight uint64, startHash common.Hash) error {
```

Comment:
Please add comment to explain what this func does.

Reply:
comments added!